src/bootstrap/download.rs

```diff
@@ -320,25 +320,42 @@ impl Config {
     }
 
     /// Returns whether the SHA256 checksum of `path` matches `expected`.
-    fn verify(&self, path: &Path, expected: &str) -> bool {
+    pub(crate) fn verify(&self, path: &Path, expected: &str) -> bool {
         use sha2::Digest;
 
         self.verbose(&format!("verifying {}", path.display()));
+
+        if self.dry_run() {
+            return false;
+        }
+
         let mut hasher = sha2::Sha256::new();
-        // FIXME: this is ok for rustfmt (4.1 MB large at time of writing), but it seems memory-intensive for rustc and larger components.
-        // Consider using streaming IO instead?
-        let contents = if self.dry_run() { vec![] } else { t!(fs::read(path)) };
-        hasher.update(&contents);
-        let found = hex::encode(hasher.finalize().as_slice());
-        let verified = found == expected;
-        if !verified && !self.dry_run() {
+
+        let file = t!(File::open(path));
+        let mut reader = BufReader::new(file);
+        let mut buffer = [0; 4096]; // read in chunks of 4KB
+
+        loop {
+            let read_len = t!(reader.read(&mut buffer));
```

Comment:
You can avoid the extra copy by using `BufReaders` `BufRead::{consume, fill_buf}` methods.
Though the overhead is probably negligible.

Reply:
Isn't that what `read` already does?

Reply:
Actually, just skip the buf-reader I guess if you're already providing a buffer.

Reply:
One of the main idea here is to iterate through the file content in small chunks of bytes, so that we don't load the entire buffer into memory all at once when reading it. So I have to provide limited(in size) buffer in each call on `BufRead::read`(which also does the `consume` from the reader automagically on each call).

Reply:
`BufReader` already has an internal buffer. So one buffer should do.

Reply:
> BufReader already has an internal buffer. So one buffer should do.

I guess you mean `File` internal buffer? If so, I switched that one, the implementation relies on the host OS, but I guess it doesn't really matter that much.

Reply:
No, `std::fs::File` does not have any userspace buffering.  `std::io::Bufreader` [does](https://doc.rust-lang.org/nightly/src/std/io/buffered/bufreader.rs.html#51)

Reply:
Ah I see the point, big thanks!